### PR TITLE
Add Sibyl Q&A to inline help

### PR DESCRIPTION
**Pull request by notnownikki**
Integrates the sibyl suggested questions with the compact view of the contact form.

Test 1:

Open the inline help.
Click "Contact Us".
Enter "domains" in the text field. The button should change to "Continue"
Click "Continue"
You should see a list of suggested questions about domains.
Click "Back"
Click "Contact Us"
The contact form should be showing the form again.

Test 2:

Go to /help/contact
Enter "domains"
Suggested questions should pop up as normal

### CODE REVIEW

File: client/state/inline-help/reducer.js [hunk 1]
```diff
@@ -79,4 +81,22 @@ export const search = createReducer(
...
 );
 
-export default combineReducers( { requesting, search } );
+export const contactForm = createReducer(
+	{
+		isShowingQandASuggestions: false,
+	},
+	{
+		[ INLINE_HELP_CONTACT_FORM_RESET ]: state => {
+			return Object.assign( {}, state, {
```
Comment (spen) [inline client/state/inline-help/reducer.js]: We can benefit from both the implicit return of the arrow function, and the spread operator here for brevity:
```js
[ INLINE_HELP_CONTACT_FORM_RESET ]: state => ( {
  ...state, 
  isShowingQandASuggestions: false,
} ),
[ INLINE_HELP_CONTACT_FORM_SHOW_QANDA ]: state => ( { 
  ...
} )

File: client/state/inline-help/selectors.js [hunk 2]
```diff
@@ -73,3 +73,13 @@ export function getInlineHelpCurrentlySelectedLink( state ) {
...
 }
+
+/**
+ * Returns a bool indicating if the contact form UI is showing the Q&A suggestions.
+ *
+ * @param  {Object}  state  Global state tree
+ * @return {Boolean}        Is the contact form UI showing the questions
+ */
+export function isShowingQandAInlineHelpContactForm( state ) {
+	return get( state, 'inlineHelpSearchResults.contactForm.isShowingQandASuggestions' );
```
Comment (spen) [inline client/state/inline-help/selectors.js]: Without a default value get will return `undefined` when `isShowingQandASuggestions` is missing and thus violate the `@return {Boolean}` doc statement.
We can use `get`s third argument to supply a default value of `true` or `false` to keep inline with the jsdocs here though.

**Comment by spen** [comment]
Some thoughts from my run through the test plan (not necessarily in scope of this work, but worth considering) :

- Updates to the contact form in the `inline-help` form are reflected in the `/help/contact/` form too. This seems a little weird and could be confusing to the customer. We might want to consider not sharing the same state slice for both forms and instead pass state through the component as props in a follow up PR (See images below for this flow)
- The copy: "Do you want the answer to any of these questions?" Feels a bit abrupt/unfriendly compared to our usual tone.
- The transition between showing and not showing also feels a bit abrupt.


Flow showing shared state:
screen shot 2018-05-01 at 15 33 34
screen shot 2018-05-01 at 15 34 27
screen shot 2018-05-01 at 15 34 50
screen shot 2018-05-01 at 15 34 58

**Comment by notnownikki** [comment]
> Updates to the contact form in the inline-help form are reflected in the /help/contact/ form too. This seems a little weird and could be confusing to the customer. We might want to consider not sharing the same state slice for both forms and instead pass state through the component as props in a follow up PR 

This is my lack of React knowledge showing up here for this. I can see the component's state is shared but I can't see how to separate it. I'd say they should be separate, but at the moment I can't see how they're shared!

> The copy: "Do you want the answer to any of these questions?" Feels a bit abrupt/unfriendly compared to our usual tone.

Suggestions welcome :smile: 

> The transition between showing and not showing also feels a bit abrupt.

I'm not sure in which sense you mean. The flip between the two UI states? Perhaps there should be some sliding animation or something? Very much open to suggestions here :smile:

File: client/me/help/help-contact-form/index.jsx [hunk 3]
```diff
@@ -298,7 +298,13 @@ export class HelpContactForm extends React.PureComponent {
 		return (
 			<ul>
 				{ this.state.qanda.map( link => {
-					return <InlineHelpCompactResult helpLink={ link } onClick={ this.trackSibylClick } />;
```
Comment (spen) [inline client/me/help/help-contact-form/index.jsx]: Well spotted 👍

File: client/me/help/help-contact-form/index.jsx [hunk 4]
```diff
@@ -322,7 +322,7 @@ export class HelpContactForm extends React.PureComponent {
 			translate,
 			showingQandAStep,
 		} = this.props;
-		const hasQASuggestions = ! isEmpty( this.state.qanda );
+		const hasQASuggestions = isArray( this.state.qanda ) && ! isEmpty( this.state.qanda );
```
Comment (spen) [inline client/me/help/help-contact-form/index.jsx]: From what I understand `this.state.qanda` should always be an array but sometimes we receive an error string and in the response in `#doQandASearch` we set `qanda` to that error string?

I think we should instead look at handling this error string differently, making `qanda`'s type strictly an `array`. This just makes for a reliable and representative piece of state and avoids having to do type checking every time we access it.

We could do that by doing something like so:
```js
...
.getQandA( query, site )
  .then( response =>
    if ( isArray( response ) ) {
      return this.setState( {
        qanda: response,
        // comments...
        sibylClicked: this.state.sibylClicked && areSameQuestions( this.state.qanda, qanda ),
      } )
    }
    // If response is not an array, empty the qanda array and set the error to the response value
    // note, we would need to handle clearing `qandaError` on successful calls.
    this.setState( { qanda: [], sibylClicked: false, qandaError:  response } )
  )
  .catch( () => this.setState( { qanda: [], sibylClicked: false } ) );
```

File: client/me/help/help-contact-form/index.jsx [hunk 5]
```diff
@@ -300,7 +326,10 @@ export class HelpContactForm extends React.PureComponent {
 			showQASuggestions,
 			showHelpLanguagePrompt,
 			translate,
+			showingQandAStep,
 		} = this.props;
+		const hasQASuggestions = isArray( this.state.qanda ) && ! isEmpty( this.state.qanda );
```
Comment (spen) [inline client/me/help/help-contact-form/index.jsx]: From what I understand `this.state.qanda` should always be an array but sometimes we receive an error string and in the response in `#doQandASearch` we set `qanda` to that error string?

I think we should instead look at handling this error string differently, making `qanda`'s type strictly an `array`. This just makes for a reliable and representative piece of state and avoids having to do type checking every time we access it.

We could do that by doing something like so:
```js
...
.getQandA( query, site )
  .then( response =>
    if ( isArray( response ) ) {
      return this.setState( {
        qanda: response,
        // comments...
        sibylClicked: this.state.sibylClicked && areSameQuestions( this.state.qanda, qanda ),
      } )
    }
    // If response is not an array, empty the qanda array and set the error to the response value
    // note, we would need to handle clearing `qandaError` on successful calls.
    this.setState( { qanda: [], sibylClicked: false, qandaError:  response } )
  )
  .catch( () => this.setState( { qanda: [], sibylClicked: false } ) );
```

Comment (notnownikki) [reply client/me/help/help-contact-form/index.jsx]: Agreed. Changing it round now!

Comment (spen) [reply client/me/help/help-contact-form/index.jsx]: Also note that if we don't actually use the `qandaError` error string anywhere we can just ignore this and not add it to state.

Comment (notnownikki) [reply client/me/help/help-contact-form/index.jsx]: Yep, that's exactly what I was going to do :smile:

Comment (spen) [reply client/me/help/help-contact-form/index.jsx]: Awesome thanks heaps! 👌

**Comment by notnownikki** [comment]
> If I remove 'domains' and wind up with no input, 'Sibyl mode' will still be active, loaded with the 'domains' result.

While that's not the end result for me (the button does revert to the non-Sibyl state) there is a delay while it makes the API call. I'm pretty sure we should skip that API call if there's no text. I'll follow up with another PR.

**Pull request merged**